README.md
# Mongoosastic
[![Build Status](https://travis-ci.org/mongoosastic/mongoosastic.svg?branch=master)](https://travis-ci.org/mongoosastic/mongoosastic)
[![NPM version](https://img.shields.io/npm/v/mongoosastic.svg)](https://www.npmjs.com/package/mongoosastic)
[![Coverage Status](https://coveralls.io/repos/mongoosastic/mongoosastic/badge.svg?branch=master&service=github)](https://coveralls.io/github/mongoosastic/mongoosastic?branch=master)
[![Downloads](https://img.shields.io/npm/dm/mongoosastic.svg)](https://www.npmjs.com/package/mongoosastic)
[![Gitter](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/mongoosastic/mongoosastic?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge)

[![NPM](https://nodei.co/npm/mongoosastic.png)](https://nodei.co/npm/mongoosastic/)


This is a fork of Mongoosastic that enables you to pass a 'parentKey' as an option to the mongoosastic plugin so that [parent->children](https://www.elastic.co/guide/en/elasticsearch/guide/current/indexing-parent-child.html) are possible.

Mongoosastic is a [mongoose](http://mongoosejs.com/) plugin that can automatically index your models into [elasticsearch](https://www.elastic.co/).

- [Installation](#installation)
- [Setup](#setup)
- [Indexing](#indexing)
  - [Saving a document](#saving-a-document)
  - [Removing a document](#removing-a-document)
  - [Indexing nested models](#indexing-nested-models)
  - [Indexing mongoose references](#indexing-mongoose-references)
  - [Indexing an existing collection](#indexing-an-existing-collection)
  - [Bulk indexing](#bulk-indexing)
  - [Filtered indexing](#filtered-indexing)
  - [Indexing on demand](#indexing-on-demand)
  - [Unindexing on demand](#unindexing-on-demand)
  - [Truncating an index](#truncating-an-index)
  - [Restrictions](#restrictions)
    - [Auto indexing](#auto-indexing)
    - [Search immediately after es-indexed event](#search-immediately-after-es-indexed-event)
- [Mapping](#mapping)
  - [Geo mapping](#geo-mapping)
    - [Indexing a geo point](#indexing-a-geo-point)
    - [Indexing a geo shape](#indexing-a-geo-shape)
  - [Creating mappings on-demand](#creating-mappings-on-demand)
- [Queries](#queries)
  - [Hydration](#hydration)

## Installation

The latest version of this package will be as close as possible to the latest `elasticsearch` and `mongoose` packages. If you are working with latest mongoose package, install normally:

```bash
npm install -S mongoosastic
```


If you are working with `mongoose@3.8.x` use `mongoosastic@2.x` and install a specific version:

```bash
npm install -S mongoosastic@^2.x
```


## Setup

### Model.plugin(mongoosastic, options)

Options are:

* `index` - the index in Elasticsearch to use. Defaults to the pluralization of the model name.
* `type`  - the type this model represents in Elasticsearch. Defaults to the model name.
* `esClient` - an existing Elasticsearch `Client` instance.
* `hosts` - an array hosts Elasticsearch is running on.
* `host` - the host Elasticsearch is running on
* `port` - the port Elasticsearch is running on
* `auth` - the authentication needed to reach Elasticsearch server. In the standard format of 'username:password'
* `protocol` - the protocol the Elasticsearch server uses. Defaults to http
* `hydrate` - whether or not to lookup results in mongodb before
* `hydrateOptions` - options to pass into hydrate function
* `bulk` - size and delay options for bulk indexing
* `filter` - the function used for filtered indexing
* `transform` - the function used to transform serialized document before indexing
* `populate` - an Array of Mongoose populate options objects
* `indexAutomatically` - allows indexing after model save to be disabled for when you need finer control over when documents are indexed. Defaults to true
* `customProperties` - an object detailing additional properties which will be merged onto the type's default mapping when `createMapping` is called.
* `saveOnSynchronize` - triggers Mongoose save (and pre-save) method when synchronizing a collection/index. Defaults to true


To have a model indexed into Elasticsearch simply add the plugin.

```javascript
var mongoose     = require('mongoose')
  , mongoosastic = require('mongoosastic')
  , Schema       = mongoose.Schema

var User = new Schema({
    name: String
  , email: String
  , city: String
})

User.plugin(mongoosastic)
```

This will by default simply use the pluralization of the model name as the index
while using the model name itself as the type. So if you create a new
User object and save it, you can see it by navigating to
http://localhost:9200/users/user/_search (this assumes Elasticsearch is
running locally on port 9200).

The default behavior is all fields get indexed into Elasticsearch. This can be a little wasteful especially considering that
the document is now just being duplicated between mongodb and
Elasticsearch so you should consider opting to index only certain fields by specifying `es_indexed` on the
fields you want to store:


```javascript
var User = new Schema({
    name: {type:String, es_indexed:true}
  , email: String
  , city: String
})

User.plugin(mongoosastic)
```

In this case only the name field will be indexed for searching.

Now, by adding the plugin, the model will have a new method called
`search` which can be used to make simple to complex searches. The `search`
method accepts [standard Elasticsearch query DSL](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-queries.html)

```javascript
User.search({
  query_string: {
    query: "john"
  }
}, function(err, results) {
  // results here
});

```

To connect to more than one host, you can use an array of hosts.

```javascript
MyModel.plugin(mongoosastic, {
  hosts: [
    'localhost:9200',
    'anotherhost:9200'
  ]
})
```

Also, you can re-use an existing Elasticsearch `Client` instance

```javascript
var esClient = new elasticsearch.Client({host: 'localhost:9200'});
MyModel.plugin(mongoosastic, {
  esClient: esClient
})
```


## Indexing

### Saving a document
The indexing takes place after saving inside the mongodb and is a defered process.
One can check the end of the indexion catching es-indexed event.


```javascript
doc.save(function(err){
  if (err) throw err;
  /* Document indexation on going */
  doc.on('es-indexed', function(err, res){
    if (err) throw err;
    /* Document is indexed */
    });
  });
```

### Removing a document
Removing a document, or unindexing, takes place when a document is removed by calling `.remove()` on a mongoose Document instance.
One can check the end of the unindexing by catching the es-removed event.

```javascript
doc.remove(function(err) {
  if (err) throw err;
  /* Document unindexing in the background */
  doc.on('es-removed', function(err, res) {
    if (err) throw err;
    /* Docuemnt is unindexed */
  });
});
```

Note that use of `Model.remove` does not involve mongoose documents as outlined in the [documentation](http://mongoosejs.com/docs/api.html#model_Model.remove). Therefore, the following will not unindex the document.

```javascript
MyModel.remove({ _id: doc.id }, function(err) {
  /* doc remains in Elasticsearch cluster */
});
```

### Indexing Nested Models
In order to index nested models you can refer following example.

```javascript
var Comment = new Schema({
    title: String
  , body: String
  , author: String
})


var User = new Schema({
    name: {type:String, es_indexed:true}
  , email: String
  , city: String
  , comments: {type:[Comment], es_indexed:true}
})

User.plugin(mongoosastic)
```

### Elasticsearch [Nested datatype](https://www.elastic.co/guide/en/elasticsearch/reference/current/nested.html)
Since the default in Elasticsearch is to take arrays and flatten them into objects,
it can make it hard to write queries where you need to maintain the relationships
between objects in the array, per .
The way to change this behavior is by changing the Elasticsearch type from `object`
(the mongoosastic default) to `nested`

```javascript
var Comment = new Schema({
    title: String
  , body: String
  , author: String
})


var User = new Schema({
    name: {type: String, es_indexed: true}
  , email: String
  , city: String
  , comments: {
      type:[Comment],
      es_indexed: true,
      es_type: 'nested',
      es_include_in_parent: true
  }
})

User.plugin(mongoosastic)
```

### Indexing Mongoose References
In order to index mongoose references you can refer following example.

```javascript
var Comment = new Schema({
    title: String
  , body: String
  , author: String
});


var User = new Schema({
    name: {type:String, es_indexed:true}
  , email: String
  , city: String
  , comments: {type: Schema.Types.ObjectId, ref: 'Comment',
    es_schema: Comment, es_indexed:true, es_select: 'title body'}
})

User.plugin(mongoosastic, {
  populate: [
    {path: 'comments', select: 'title body'}
  ]
})
```
In the schema you'll need to provide `es_schema` field - the referenced schema.
By default every field of the referenced schema will be mapped. Use `es_select` field to pick just specific fields.

`populate` is an array of options objects you normally pass to
[Model.populate](http://mongoosejs.com/docs/api.html#model_Model.populate).

### Indexing An Existing Collection
Already have a mongodb collection that you'd like to index using this
plugin? No problem! Simply call the synchronize method on your model to
open a mongoose stream and start indexing documents individually.

```javascript
var BookSchema = new Schema({
  title: String
});
BookSchema.plugin(mongoosastic);

var Book = mongoose.model('Book', BookSchema)
  , stream = Book.synchronize()
  , count = 0;

stream.on('data', function(err, doc){
  count++;
});
stream.on('close', function(){
  console.log('indexed ' + count + ' documents!');
});
stream.on('error', function(err){
  console.log(err);
});
```

You can also synchronize a subset of documents based on a query!

```javascript
var stream = Book.synchronize({author: 'Arthur C. Clarke'})
```

As well as specifying synchronization options

```javascript
var stream = Book.synchronize({}, {saveOnSynchronize: true})
```

Options are:

 * `saveOnSynchronize` - triggers Mongoose save (and pre-save) method when synchronizing a collection/index. Defaults to global `saveOnSynchronize` option


### Bulk Indexing

You can also specify `bulk` options with mongoose which will utilize Elasticsearch's bulk indexing api. This will cause the `synchronize` function to use bulk indexing as well.

Mongoosastic will wait 1 second (or specified delay) until it has 1000 docs (or specified size) and then perform bulk indexing.

```javascript
BookSchema.plugin(mongoosastic, {
  bulk: {
    size: 10, // preferred number of docs to bulk index
    delay: 100 //milliseconds to wait for enough docs to meet size constraint
  }
});
```

### Filtered Indexing

You can specify a filter function to index a model to Elasticsearch based on some specific conditions.

Filtering function must return True for conditions that will ignore indexing to Elasticsearch.

```javascript
var MovieSchema = new Schema({
  title: {type: String},
  genre: {type: String, enum: ['horror', 'action', 'adventure', 'other']}
});

MovieSchema.plugin(mongoosastic, {
  filter: function(doc) {
    return doc.genre === 'action';
  }
});
```

Instances of Movie model having 'action' as their genre will not be indexed to Elasticsearch.


### Indexing On Demand
You can do on-demand indexes using the `index` function

```javascript
Dude.findOne({name:'Jeffrey Lebowski', function(err, dude){
  dude.awesome = true;
  dude.index(function(err, res){
    console.log("egads! I've been indexed!");
  });
});
```

The index method takes 2 arguments:


CHANGELOG.md
4.4.1.2 2018-05-17
==========
  * rebased with upstream

4.3.2 2017-03-??
==========
  * add the ability to use the parentKey to index different documents "together". Thus, we can get all documents in one scoop

4.3.0 2017-02-21
==========
  * chore(package): update eslint to version 3.16.0
    https://greenkeeper.io/
  * chore(package): update mongoose to version 4.8.4
    https://greenkeeper.io/
  * chore(package): update async to version 2.1.5
    https://greenkeeper.io/
  * chore(package): update eslint-plugin-promise to version 3.4.2
    https://greenkeeper.io/
  * chore(package): update mongoose to version 4.8.3
    https://greenkeeper.io/
  * chore(package): update standard to version 9.0.0-beta.0 ([#292](https://github.com/mongoosastic/mongoosastic/issues/292))
    https://greenkeeper.io/
  * chore(package): update mongoose to version 4.8.2 ([#293](https://github.com/mongoosastic/mongoosastic/issues/293))
    https://greenkeeper.io/
  * chore(package): update eslint-plugin-promise to version 3.4.1
    https://greenkeeper.io/
  * Fix mapping issue for id.
    Create mocha test for mapping id
  * chore(package): update coveralls to version 2.11.16
    https://greenkeeper.io/
  * chore(package): update mongoose to version 4.8.1 ([#286](https://github.com/mongoosastic/mongoosastic/issues/286))
    Closes [#281](https://github.com/mongoosastic/mongoosastic/issues/281)
    https://greenkeeper.io/
  * chore(package): update eslint to version 3.15.0 ([#285](https://github.com/mongoosastic/mongoosastic/issues/285))
    https://greenkeeper.io/
  * chore(package): update mongoose to version 4.8.0
    https://greenkeeper.io/
  * chore(package): update mongoose to version 4.7.9
    https://greenkeeper.io/
  * chore(package): update eslint to version 3.14.1 ([#279](https://github.com/mongoosastic/mongoosastic/issues/279))
    https://greenkeeper.io/
  * chore(package): update mongoose to version 4.7.7 ([#273](https://github.com/mongoosastic/mongoosastic/issues/273))
    https://greenkeeper.io/
  * docs: createMapping should called manualy
  * chore(package): update eslint to version 3.13.1
    https://greenkeeper.io/
  * add test for refresh
  * update coverage when tests passed; use mocha.opts
  * fix for stable test results

4.2.4 2017-01-06
==========

  * chore(package): update eslint to version 3.13.0 ([#267](https://github.com/mongoosastic/mongoosastic/issues/267))
    https://greenkeeper.io/
  * chore(package): update dependencies ([#266](https://github.com/mongoosastic/mongoosastic/issues/266))
    https://greenkeeper.io/
  * resolve [#95](https://github.com/mongoosastic/mongoosastic/issues/95), support routing on index/unIndex/esSearch. ([#260](https://github.com/mongoosastic/mongoosastic/issues/260))
    * support routing on index/unIndex/esSearch.
  * update travis build badge to precise one ([#265](https://github.com/mongoosastic/mongoosastic/issues/265))
  * support auto index when call Model.insertMany
  * add co-mocha: write tests with generator function
  * fix bad request when esCount without query ([#261](https://github.com/mongoosastic/mongoosastic/issues/261))
  * unIndex should retry when unindex failed ([#259](https://github.com/mongoosastic/mongoosastic/issues/259))
  * unIndex should retry when unindex failed
  * should trigger removed event when retry failed
  * fix [#111](https://github.com/mongoosastic/mongoosastic/issues/111), findOneAndRemove should work fine when no document found ([#258](https://github.com/mongoosastic/mongoosastic/issues/258))
  * add *.swp to .gitignore
  * Merge pull request [#257](https://github.com/mongoosastic/mongoosastic/issues/257) from mongoosastic/issue/docs
    update elasticsearch.org to elastic.co
  * update elasticsearch.org to elastic.co

4.2.2 2016-12-22
==========

  * test folder linted ([#247](https://github.com/mongoosastic/mongoosastic/issues/247))
  * test folder linted
  * Exclude aggs from being added to the esQuery, as they're already added to the body. Fixes errors with Elasticsearch v5.0.0.
  * added a catch to the populate promise call in postSave ([#248](https://github.com/mongoosastic/mongoosastic/issues/248))
  * should keep populated field of source doc unchanged ([#254](https://github.com/mongoosastic/mongoosastic/issues/254))
  * should keep populated field of source doc unchanged
  * fix typo

2016-12-12
==========

  * Fix ([#5](https://github.com/mongoosastic/mongoosastic/issues/5)) mapping creation of nested references ([#252](https://github.com/mongoosastic/mongoosastic/issues/252))
  * Update mapping-generator.js
  * Test proving that mongoosastic fails when mapping a geo_point field inside a nested referenced schema
  * Restoring original mapping-generator.js file
  * Tests fixed and array of nested referenced object now correctly mapped


4.2.1 2016-11-20
==========

  * update dependencies ([#246](https://github.com/mongoosastic/mongoosastic/issues/246))
  * eslint updated ([#245](https://github.com/mongoosastic/mongoosastic/issues/245))
    * eslint updated
    * dependencies updated & added strict mode

4.1.0 2016-11-17
==========

  * Fixed: [#207](https://github.com/mongoosastic/mongoosastic/issues/207) Add min_score to search function ([#218](https://github.com/mongoosastic/mongoosastic/issues/218))
  * Reset bulk to original options.bulk instead of false. Fixes [#161](https://github.com/mongoosastic/mongoosastic/issues/161) ([#237](https://github.com/mongoosastic/mongoosastic/issues/237))
  * fix a bug where calling doc.index would overwrite all plugin options! ([#187](https://github.com/mongoosastic/mongoosastic/issues/187))
  * Custom serializer ([#241](https://github.com/mongoosastic/mongoosastic/issues/241))
    * updated the index method to allow for specifying a custom serialize method
    * added test for the custom serializer

lib/mongoosastic.js
'use strict'

const elasticsearch = require('elasticsearch')
const Generator = require('./mapping-generator')
const generator = new Generator()
const serialize = require('./serialize')
const events = require('events')
const util = require('util')
const nop = function nop () {}

function isString (subject) {
  return typeof subject === 'string'
}

function isStringArray (arr) {
  return arr.filter && arr.length === (arr.filter(item => typeof item === 'string')).length
}

function getMapping (schema) {
  let retMapping = {}
  generator.generateMapping(schema, (err, mapping) => {
    if (err) {
      // no handling needed
    }
    retMapping = mapping
  })

  return retMapping
}

function createEsClient (options) {
  const esOptions = {}

  if (util.isArray(options.hosts)) {
    esOptions.host = options.hosts
  } else {
    esOptions.host = {
      host: options && options.host ? options.host : 'localhost',
      port: options && options.port ? options.port : 9200,
      protocol: options && options.protocol ? options.protocol : 'http',
      auth: options && options.auth ? options.auth : null,
      keepAlive: false
    }
  }

  esOptions.log = (options ? options.log : null)

  return new elasticsearch.Client(esOptions)
}

function createMappingIfNotPresent (options, cb) {
  const client = options.client
  const indexName = options.indexName
  const typeName = options.typeName
  const schema = options.schema
  const settings = options.settings
  const properties = options.properties

  generator.generateMapping(schema, (ignoredErr, mapping) => {
    const completeMapping = {}
    completeMapping[typeName] = mapping

    if (properties) {
      Object.keys(properties).map(key => {
        completeMapping[typeName].properties[key] = properties[key]
      })
    }

    client.indices.exists({
      index: indexName
    }, (err, exists) => {
      if (err) {
        return cb(err)
      }

      if (exists) {
        return client.indices.putMapping({
          index: indexName,
          type: typeName,
          body: completeMapping
        }, cb)
      }
      return client.indices.create({
        index: indexName,
        body: settings
      }, indexErr => {
        if (indexErr) {
          return cb(indexErr)
        }

        client.indices.putMapping({
          index: indexName,
          type: typeName,
          body: completeMapping
        }, cb)
      })
    })
  })
}

function hydrate (res, model, options, cb) {
  const results = res.hits
  const resultsMap = {}
  const ids = results.hits.map((result, idx) => {
    resultsMap[result._id] = idx
    return result._id
  })

  const query = model.find({
    _id: {
      $in: ids
    }
  })
  const hydrateOptions = options.hydrateOptions

  // Build Mongoose query based on hydrate options
  // Example: {lean: true, sort: '-name', select: 'address name'}
  Object.keys(hydrateOptions).forEach(option => {
    query[option](hydrateOptions[option])
  })

  query.exec((err, docs) => {
    let hits
    let docsMap = {}

    if (err) {
      return cb(err)
    }

    if (!docs || docs.length === 0) {
      results.hits = []
      res.hits = results
      return cb(null, res)
    }

    if (hydrateOptions.sort) {
      // Hydrate sort has precedence over ES result order
      hits = docs
    } else {
      // Preserve ES result ordering
      docs.forEach(doc => {
        docsMap[doc._id] = doc
      })
      hits = results.hits.map(result => docsMap[result._id])
    }

    if (options.highlight || options.hydrateWithESResults) {
      hits.forEach(doc => {
        const idx = resultsMap[doc._id]
        if (options.highlight) {
          doc._highlight = results.hits[idx].highlight
        }
        if (options.hydrateWithESResults) {
          // Add to doc ES raw result (with, e.g., _score value)
          doc._esResult = results.hits[idx]
          if (!options.hydrateWithESResults.source) {
            // Remove heavy load
            delete doc._esResult._source
          }
        }
      })
    }

    results.hits = hits
    res.hits = results
    cb(null, res)
  })
}

function deleteByMongoId (options, cb) {
  const index = options.index
  const type = options.type
  const client = options.client
  const model = options.model
  const routing = options.routing
  let tries = options.tries

  client.delete({
    index: index,
    type: type,
    id: model._id.toString(),
    routing: routing
  }, (err, res) => {
    if (err && err.status === 404) {
      if (tries <= 0) {
        model.emit('es-removed', err, res)
        return cb(err)
      }
      options.tries = --tries
      setTimeout(() => {
        deleteByMongoId(options, cb)
      }, 500)
    } else {
      model.emit('es-removed', err, res)
      cb(err)
    }
  })
}

function Mongoosastic (schema, pluginOpts) {
  let options = pluginOpts || {}

  let bulkTimeout
  let bulkBuffer = []
  let esClient
  let populate = options && options.populate
  let mapping = getMapping(schema)

  let indexName = options && options.index
  let typeName = options && options.type
  let alwaysHydrate = options && options.hydrate
  let defaultHydrateOptions = options && options.hydrateOptions
  let bulk = options && options.bulk
  let filter = options && options.filter
  let transform = options && options.transform
  let routing = options && options.routing

  let customProperties = options && options.customProperties
  let customSerialize = options && options.customSerialize
  let indexAutomatically = !(options && options.indexAutomatically === false)
  let saveOnSynchronize = !(options && options.saveOnSynchronize === false)
  let parentKey = (options && options.parentKey) || false
  let forceIndexRefresh = !(options && options.forceIndexRefresh === false)

  if (options.esClient) {
    esClient = options.esClient
  } else {
    esClient = createEsClient(options)
  }

  function setIndexNameIfUnset (model) {
    const modelName = model.toLowerCase()
    if (!indexName) {
      indexName = `${modelName}s`
    }

    if (!typeName) {
      typeName = modelName
    }
  }

  function postSave (doc) {
    let _doc
    function onIndex (err, res) {
      if (!filter || !filter(doc)) {
        doc.emit('es-indexed', err, res)
      } else {
        doc.emit('es-filtered', err, res)
      }
    }

    if (doc) {
      _doc = new doc.constructor(doc)
      if (populate && populate.length) {
        populate.forEach(populateOpts => {
          _doc.populate(populateOpts)
        })
        _doc.execPopulate().then(popDoc => {
          popDoc.index(onIndex)
        }).catch(onIndex)
      } else {
        _doc.index(onIndex)
      }
    }
  }

  function clearBulkTimeout () {
    clearTimeout(bulkTimeout)
    bulkTimeout = undefined
  }

  function bulkAdd (instruction) {
    bulkBuffer.push(instruction)

    // Return because we need the doc being indexed
    // Before we start inserting
    if (instruction.index && instruction.index._index) {
      return
    }

    if (bulkBuffer.length >= ((bulk && bulk.size) || 1000)) {
      schema.statics.flush()
      clearBulkTimeout()
    } else if (bulkTimeout === undefined) {
      bulkTimeout = setTimeout(() => {
        schema.statics.flush()
        clearBulkTimeout()
      }, (bulk && bulk.delay) || 1000)
    }
  }

  function bulkDelete(opts, cb) {
    let inst = {
	    delete: {
		    _index: opts.index || indexName,
		    _type: opts.type || typeName,
		    _id: opts.model._id.toString(),
		    _routing: opts.routing
	    }
    };
    if (parentKey && opts.model[parentKey]) {
	    inst.delete.parent = opts.model[parentKey].toString()
    }
    bulkAdd(inst);
    cb();
  }

  function bulkIndex(opts) {
    let inst = {
	    index: {
		    _index: opts.index || indexName,
		    _type: opts.type || typeName,
		    _id: opts._id.toString(),
		    _routing: opts.routing
	    }
    };
    if (parentKey && opts.model[parentKey]) {
	    inst.index.parent = opts.model[parentKey].toString()

	    // console.log("mongoosastic: parentKey",parentKey," is ", inst.index.parent);
    }

	bulkAdd(inst);
	bulkAdd(opts.model);
  }

  /**
   * ElasticSearch Client
   */
  schema.statics.esClient = esClient

  /**
   * Create the mapping. Takes an optional settings parameter
   * and a callback that will be called once the mapping is created

   * @param settings Object (optional)
   * @param cb Function
   */
  schema.statics.createMapping = function createMapping (inSettings, inCb) {
    let cb = inCb
    let settings = inSettings
    if (arguments.length < 2) {
      cb = inSettings || nop
      settings = undefined
    }

    setIndexNameIfUnset(this.modelName)

    createMappingIfNotPresent({
      client: esClient,
      indexName: indexName,
      typeName: typeName,
      schema: schema,
      settings: settings,
      properties: customProperties
    }, cb)
  }

  /**
   * @param options  Object (optional)
   * @param cb Function
   */
  schema.methods.index = function schemaIndex (inOpts, inCb) {
    let index
    let type
    let serialModel
    let cb = inCb
    let opts = inOpts
    let _serialize = serialize

    if (arguments.length < 2) {
      cb = inOpts || nop
      opts = {}
    }

    if (filter && filter(this)) {
      return this.unIndex(cb)
    }

    if (typeof customSerialize === 'function') {
      _serialize = customSerialize
    }

    setIndexNameIfUnset(this.constructor.modelName)

    index = opts.index || indexName
    type = opts.type || typeName

    /**
     * Serialize the model, and apply transformation
     */
    serialModel = _serialize(this, mapping)
    if (transform) serialModel = transform(serialModel, this)

    const _opts = {
      index: index,
      type: type,
      refresh: forceIndexRefresh
    }
    if (routing) {
        _opts.routing = routing(this)
    }

    if (bulk) {
	    _opts.model = serialModel
	    _opts._id = this._id
	    bulkIndex(_opts)
	    setImmediate(() => cb(null, this))
    } else {
      _opts.id = this._id.toString()
      _opts.body = serialModel
      if (parentKey && this[parentKey]) {
	      _opts.parent = this[parentKey].toString();
      }
      esClient.index(_opts, cb)
    }
  }

  /**
   * Unset elasticsearch index
   * @param options - (optional) options for unIndex
   * @param cb - callback when unIndex is complete
   */
  schema.methods.unIndex = function unIndex (inOpts, inCb) {
    let opts = inOpts
    let cb = inCb

    if (arguments.length < 2) {
      cb = inOpts || nop
      opts = {}
    }

    setIndexNameIfUnset(this.constructor.modelName)

    opts.index = opts.index || indexName
    opts.type = opts.type || typeName
    opts.model = this
    opts.client = esClient
    opts.tries = opts.tries || 3
    if (routing) {
      opts.routing = routing(this)
    }

    if (parentKey && opts.model[parentKey]) {
	    opts.parent = opts.model[parentKey].toString()
    }


    if (bulk) {
      bulkDelete(opts, cb)
    } else {
      deleteByMongoId(opts, cb)
    }
  }

  /**
   * Delete all documents from a type/index
   * @param options - (optional) specify index/type
   * @param cb - callback when truncation is complete
   */
  schema.statics.esTruncate = function esTruncate (inOpts, inCb) {
    let opts = inOpts
    let cb = inCb
    let esQuery

    if (arguments.length < 2) {
      cb = inOpts || nop
      opts = {}
    }

    setIndexNameIfUnset(this.modelName)

    opts.index = opts.index || indexName
    opts.type = opts.type || typeName

    esQuery = {
      body: {
        query: {
          match_all: {}
        }
      },
      index: opts.index,
      type: opts.type
    }

    esClient.search(esQuery, (err, res) => {
      if (err) {
        return cb(err)
      }
      if (res.hits.total) {
        res.hits.hits.forEach(doc => {
          opts.model = doc
          if (routing) {
            doc._source._id = doc._id
            opts.routing = routing(doc._source)
          }
          bulkDelete(opts, nop)
        })
      }
      cb()
    })
  }

  /**
   * Synchronize an existing collection
   *
   * @param query - query for documents you want to synchronize
   */
  schema.statics.synchronize = function synchronize (inQuery, inOpts) {
    let em = new events.EventEmitter()
    let closeValues = []
    let counter = 0
    let stream
    let query = inQuery || {}
    let close = function close () {
      em.emit.apply(em, ['close'].concat(closeValues))
    }

    let _saveOnSynchronize = inOpts &&
      inOpts.saveOnSynchronize !== undefined ? inOpts.saveOnSynchronize : saveOnSynchronize

    // Set indexing to be bulk when synchronizing to make synchronizing faster
    // Set default values when not present
    bulk = {
      delay: bulk && bulk.delay || 1000,
      size: bulk && bulk.size || 1000,
      batch: bulk && bulk.batch || 50
    }

    setIndexNameIfUnset(this.modelName)

    stream = this.find(query).batchSize(bulk.batch).cursor()

    stream.on('data', doc => {
      stream.pause()
      counter++

      function onIndex (indexErr, inDoc) {
        counter--
        if (indexErr) {
          em.emit('error', indexErr)
        } else {
          em.emit('data', null, inDoc)
        }
        stream.resume()
      }

      doc.on('es-indexed', onIndex)
      doc.on('es-filtered', onIndex)

      if (_saveOnSynchronize) {
        // Save document with Mongoose first
        doc.save(err => {
          if (err) {
            counter--
            em.emit('error', err)
            return stream.resume()
          }
        })
      } else {
        postSave(doc)
      }
    })

    stream.on('close', (pA, pB) => {
      let closeInterval
      closeValues = [pA, pB]
      closeInterval = setInterval(() => {
        if (counter === 0 && bulkBuffer.length === 0) {
          clearInterval(closeInterval)
          close()
          bulk = options && options.bulk
        }
      }, 1000)
    })

    stream.on('error', err => {
      em.emit('error', err)
    })

    return em
  }

  /**
   * ElasticSearch search function
   *
   * Wrapping schema.statics.es_search().
   *
   * @param inQuery - query object to perform search with
   * @param inOpts - (optional) special search options, such as hydrate
   * @param inCb - callback called with search results
   */
  schema.statics.search = function search (inQuery, inOpts, inCb) {
    let cb = inCb
    let opts = inOpts
    let query = inQuery === null ? undefined : inQuery
    let esSearch
    let fullQuery

    if (arguments.length === 2) {
      cb = arguments[1]
      opts = {}
    }

    fullQuery = {
      query: query
    }

    esSearch = schema.statics.esSearch.bind(this)

    return esSearch(fullQuery, opts, cb)
  }

  /**
   * ElasticSearch true/raw search function
   *
   * Elastic search query: provide full query object.
   * Useful, e.g., for paged requests.
   *
   * @param inQuery - **full** query object to perform search with
   * @param inOpts - (optional) special search options, such as hydrate
   * @param inCb - callback called with search results
     */
  schema.statics.esSearch = function (inQuery, inOpts, inCb) {
    let _this = this
    let cb = inCb
    let opts = inOpts
    let esQuery
    let query = inQuery === null ? undefined : inQuery

    if (arguments.length === 2) {
      cb = arguments[1]
      opts = {}
    }

    opts.hydrateOptions = opts.hydrateOptions || defaultHydrateOptions || {}

    setIndexNameIfUnset(this.modelName)

    esQuery = {
      body: query,
      index: opts.index || indexName,
      type: opts.type || typeName
    }

    if (opts.routing) {
      esQuery.routing = opts.routing
    }

    if (opts.highlight) {
      esQuery.body.highlight = opts.highlight
    }

    if (opts.suggest) {
      esQuery.body.suggest = opts.suggest
    }

    if (opts.aggs) {
      esQuery.body.aggs = opts.aggs
    }

    if (opts.min_score) {
      esQuery.body.min_score = opts.min_score
    }

    Object.keys(opts).forEach(opt => {
      if (!opt.match(/(hydrate|sort|aggs)/) && opts.hasOwnProperty(opt)) {
        esQuery[opt] = opts[opt]
      }

      if (opts.sort) {
        if (isString(opts.sort) || isStringArray(opts.sort)) {
          esQuery.sort = opts.sort
        } else {
          esQuery.body.sort = opts.sort
        }
      }
    })

    esClient.search(esQuery, (err, res) => {
      if (err) {
        return cb(err)
      }

      if (alwaysHydrate || opts.hydrate) {
        hydrate(res, _this, opts, cb)
      } else {
        cb(null, res)
      }
    })
  }

  schema.statics.esCount = function esCount (inQuery, inCb) {
    let cb = inCb
    let query = inQuery
    let esQuery

    setIndexNameIfUnset(this.modelName)

    if (!cb && typeof query === 'function') {
      cb = query
      query = {
        match_all: {}
      }
    }

    esQuery = {
      body: {
        query: query
      },
      index: indexName,
      type: typeName
    }

    esClient.count(esQuery, cb)
  }

  schema.statics.flush = function flush (inCb) {
    let cb = inCb || nop

    esClient.bulk({
      body: bulkBuffer
    }, cb)

    bulkBuffer = []
  }

  schema.statics.refresh = function refresh (inOpts, inCb) {
    let cb = inCb
    let opts = inOpts
    if (arguments.length < 2) {
      cb = inOpts || nop
      opts = {}
    }

    setIndexNameIfUnset(this.modelName)
    esClient.indices.refresh({
      index: opts.index || indexName
    }, cb)
  }

  function postRemove (doc) {
    if (!doc) {
      return
    }

    const opts = {
      index: indexName,
      type: typeName,
      tries: 3,
      model: doc,
      client: esClient
    }
    if (routing) {
      opts.routing = routing(doc)
    }

    setIndexNameIfUnset(doc.constructor.modelName)

    if (bulk) {
      bulkDelete(opts, nop)
    } else {
      deleteByMongoId(opts, nop)
    }
  }

  /**
   * Use standard Mongoose Middleware hooks
   * to persist to Elasticsearch
   */
  function setUpMiddlewareHooks (inSchema) {
    /**
     * Remove in elasticsearch on remove
     */
    inSchema.post('remove', postRemove)
    inSchema.post('findOneAndRemove', postRemove)

    /**
     * Save in elasticsearch on save.
     */
    inSchema.post('save', postSave)
    inSchema.post('findOneAndUpdate', postSave)
    inSchema.post('insertMany', (docs) => {
      docs.forEach((doc) => postSave(doc))
    })
  }

  if (indexAutomatically) {
    setUpMiddlewareHooks(schema)
  }
}

module.exports = Mongoosastic
